File: ios/Plugin/AudioAsset.swift
//
//  AudioAsset.swift
//  Plugin
//
//  Created by priyank on 2020-05-29.
//  Copyright © 2022 Martin Donadieu. All rights reserved.
//

import AVFoundation

public class AudioAsset: NSObject, AVAudioPlayerDelegate {

    var channels: [AVAudioPlayer] = []
    var playIndex: Int = 0
    var assetId: String = ""
    var initialVolume: Float = 1.0
    var fadeDelay: Float = 1.0
    var owner: NativeAudio

    let FADESTEP: Float = 0.05
    let FADEDELAY: Float = 0.08

    init(owner: NativeAudio, withAssetId assetId: String, withPath path: String!, withChannels channels: Int!, withVolume volume: Float!, withFadeDelay delay: Float!) {

        self.owner = owner
        self.assetId = assetId
        self.channels = []

        super.init()

        let pathUrl: URL = URL(string: path.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed)!)!
        for _ in 0..<channels {
            do {
                let player: AVAudioPlayer! = try AVAudioPlayer(contentsOf: pathUrl)

                if player != nil {
                    player.volume = volume
                    player.prepareToPlay()
                    self.channels.append(player)
                    if channels == 1 {
                        player.delegate = self
                    }
                }
            } catch {
                print("Error loading \(String(describing: path))")
            }
        }
    }

    func getCurrentTime() -> TimeInterval {
        if channels.count != 1 {
            return 0
        }
        let player: AVAudioPlayer = channels[playIndex]

        return player.currentTime
    }

    func getDuration() -> TimeInterval {
        if channels.count != 1 {
            return 0
        }

        let player: AVAudioPlayer = channels[playIndex]

        return player.duration
    }

    func play(time: TimeInterval) {
        let player: AVAudioPlayer = channels[playIndex]
        player.currentTime = time
        player.numberOfLoops = 0
        player.play()
        playIndex += 1
        playIndex = playIndex % channels.count
    }

    func playWithFade(time: TimeInterval) {
        let player: AVAudioPlayer = channels[playIndex]
        player.currentTime = time

        if !player.isPlaying {
            player.numberOfLoops = 0
            player.volume = 0
            player.play()
            playIndex += 1
            playIndex = playIndex % channels.count
        } else {
            if player.volume < initialVolume {
                player.volume += self.FADESTEP
            }
        }

    }

    func pause() {
        let player: AVAudioPlayer = channels[playIndex]
        player.pause()
    }

    func resume() {
        let player: AVAudioPlayer = channels[playIndex]

        let timeOffset = player.deviceCurrentTime + 0.01
        player.play(atTime: timeOffset)
    }

    func stop() {
        for player in channels {
            player.stop()
        }
    }

    func stopWithFade() {
        let player: AVAudioPlayer = channels[playIndex]

        if !player.isPlaying {
            player.currentTime = 0.0
            player.numberOfLoops = 0
            player.volume = 0
            player.play()
            playIndex += 1
            playIndex = playIndex % channels.count
        } else {
            if player.volume < initialVolume {
                player.volume += self.FADESTEP
            }
        }
    }

    func loop() {
        self.stop()

        let player: AVAudioPlayer = channels[playIndex]
        player.numberOfLoops = -1
        player.play()
        playIndex += 1
        playIndex = playIndex % channels.count
    }

    func unload() {
        self.stop()

        //        for i in 0..<channels.count {
        //            var player: AVAudioPlayer! = channels.object(at: i) as? AVAudioPlayer
        //
        //            player = nil
        //        }
        channels = []
    }

    func setVolume(volume: NSNumber!) {
        for player in channels {
            player.volume = volume.floatValue
        }
    }

    func setRate(rate: NSNumber!) {
        for player in channels {
            player.rate = rate.floatValue
        }
    }

    public func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully flag: Bool) {
        NSLog("playerDidFinish")
        self.owner.notifyListeners("complete", data: [
            "assetId": self.assetId
        ])
    }

    func playerDecodeError(player: AVAudioPlayer!, error: NSError!) {

    }

    func isPlaying() -> Bool {
        if channels.count != 1 {
            return false
        }

        let player: AVAudioPlayer = channels[playIndex]
        return player.isPlaying
    }
}

File: android/src/main/java/ee/forgr/audio/AudioAsset.java
package ee.forgr.audio;

import android.content.res.AssetFileDescriptor;
import android.os.Build;
import androidx.annotation.RequiresApi;
import java.util.ArrayList;

public class AudioAsset {

  private final String TAG = "AudioAsset";

  private final ArrayList<AudioDispatcher> audioList;
  private int playIndex = 0;
  private final String assetId;
  private final NativeAudio owner;

  AudioAsset(
    NativeAudio owner,
    String assetId,
    AssetFileDescriptor assetFileDescriptor,
    int audioChannelNum,
    float volume
  ) throws Exception {
    audioList = new ArrayList<>();
    this.owner = owner;
    this.assetId = assetId;

    if (audioChannelNum < 0) {
      audioChannelNum = 1;
    }

    for (int x = 0; x < audioChannelNum; x++) {
      AudioDispatcher audioDispatcher = new AudioDispatcher(
        assetFileDescriptor,
        volume
      );
      audioList.add(audioDispatcher);
      if (audioChannelNum == 1) audioDispatcher.setOwner(this);
    }
  }

  public void dispatchComplete() {
    this.owner.dispatchComplete(this.assetId);
  }

  public void play(Double time) throws Exception {
    AudioDispatcher audio = audioList.get(playIndex);

    if (audio != null) {
      audio.play(time);
      playIndex++;
      playIndex = playIndex % audioList.size();
    } else {
      throw new Exception("AudioDispatcher is null");
    }
  }

  public double getDuration() {
    if (audioList.size() != 1) return 0;

    AudioDispatcher audio = audioList.get(playIndex);

    if (audio != null) {
      return audio.getDuration();
    }
    return 0;
  }

  public double getCurrentPosition() {
    if (audioList.size() != 1) return 0;

    AudioDispatcher audio = audioList.get(playIndex);

    if (audio != null) {
      return audio.getCurrentPosition();
    }
    return 0;
  }

  public boolean pause() throws Exception {
    boolean wasPlaying = false;

    for (int x = 0; x < audioList.size(); x++) {
      AudioDispatcher audio = audioList.get(x);
      wasPlaying |= audio.pause();
    }

    return wasPlaying;
  }

  public void resume() throws Exception {
    if (!audioList.isEmpty()) {
      AudioDispatcher audio = audioList.get(0);

      if (audio != null) {
        audio.resume();
      } else {
        throw new Exception("AudioDispatcher is null");
      }
    }
  }

  public void stop() throws Exception {
    for (int x = 0; x < audioList.size(); x++) {
      AudioDispatcher audio = audioList.get(x);

      if (audio != null) {
        audio.stop();
      } else {
        throw new Exception("AudioDispatcher is null");
      }
    }
  }

  public void loop() throws Exception {
    AudioDispatcher audio = audioList.get(playIndex);

    if (audio != null) {
      audio.loop();
      playIndex++;
      playIndex = playIndex % audioList.size();
    } else {
      throw new Exception("AudioDispatcher is null");
    }
  }

  public void unload() throws Exception {
    this.stop();

    for (int x = 0; x < audioList.size(); x++) {
      AudioDispatcher audio = audioList.get(x);

      if (audio != null) {
        audio.unload();
      } else {
        throw new Exception("AudioDispatcher is null");
      }
    }

    audioList.clear();
  }

  public void setVolume(float volume) throws Exception {
    for (int x = 0; x < audioList.size(); x++) {
      AudioDispatcher audio = audioList.get(x);

      if (audio != null) {
        audio.setVolume(volume);
      } else {
        throw new Exception("AudioDispatcher is null");
      }
    }
  }

  @RequiresApi(api = Build.VERSION_CODES.M)
  public void setRate(float rate) throws Exception {
    for (int x = 0; x < audioList.size(); x++) {
      AudioDispatcher audio = audioList.get(x);
      if (audio != null) {
        audio.setRate(rate);
      }
    }
  }

  public boolean isPlaying() throws Exception {
    if (audioList.size() != 1) return false;

    return audioList.get(playIndex).isPlaying();
  }
}
